src/providers/completer/environment.ts

```diff
@@ -4,59 +4,171 @@ import {latexParser} from 'latex-utensils'
 
 import {Extension} from '../../main'
 
+export interface EnvItemEntry {
+    name: string, // Name of the environment, what comes inside \begin{...}
+    snippet?: string, // To be inserted after \begin{..}
+    package?: string, // The package providing the environment
+    detail?: string
+}
+
+export enum EnvSnippetType { AsName, AsCommand, ForBegin, }
+
+export interface Suggestion extends vscode.CompletionItem {
+    package: string
+}
+
 export class Environment {
     extension: Extension
-    private defaultEnvs: vscode.CompletionItem[] = []
-    private packageEnvs: {[pkg: string]: vscode.CompletionItem[]} = {}
+    private defaultEnvsAsName: Suggestion[] = []
+    private defaultEnvsAsCommand: Suggestion[] = []
+    private defaultEnvsForBegin: Suggestion[] = []
+    private packageEnvsAsName: {[pkg: string]: Suggestion[]} = {}
+    private packageEnvsAsCommand: {[pkg: string]: Suggestion[]} = {}
+    private packageEnvsForBegin: {[pkg: string]: Suggestion[]} = {}
 
     constructor(extension: Extension) {
         this.extension = extension
     }
 
-    initialize(envs: string[]) {
-        this.defaultEnvs = envs.map(env => new vscode.CompletionItem(env, vscode.CompletionItemKind.Module))
+    initialize(envs: {[key: string]: EnvItemEntry}) {
+        this.defaultEnvsAsCommand = []
+        this.defaultEnvsForBegin = []
+        this.defaultEnvsAsName = []
+        const endCompletion: Suggestion = {
+            label: 'Complete with \\end',
+            sortText: ' ',
+            insertText: new vscode.SnippetString('$1}\n\t$0\n\\end{$1}'),
+            command: { title: 'Post-Action', command: 'editor.action.triggerSuggest' },
+            kind: vscode.CompletionItemKind.Function,
```

Comment:
Is it intentional to use `vscode.CompletionItemKind.Function` for environments here? Previously and other cases in the same file, `vscode.CompletionItemKind.Module` is used.